components/patient-data/indexing/src/main/java/org/phenotips/data/indexing/internal/PatientEmptyObjectsRemover.java

```diff
@@ -0,0 +1,116 @@
+import org.xwiki.context.Execution;
+import org.xwiki.model.EntityType;
+import org.xwiki.model.reference.EntityReference;
+import org.xwiki.observation.AbstractEventListener;
+import org.xwiki.observation.event.Event;
+
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+
+import javax.inject.Inject;
+import javax.inject.Named;
+import javax.inject.Singleton;
+
+import org.apache.commons.lang3.StringUtils;
+
+import com.xpn.xwiki.XWikiContext;
+import com.xpn.xwiki.XWikiException;
+import com.xpn.xwiki.doc.XWikiDocument;
+import com.xpn.xwiki.objects.BaseObject;
+import com.xpn.xwiki.objects.BaseStringProperty;
+
+/**
+ * Removes gene and variant objects from the document if key fields are empty, gene field for genes and cdna field for
+ * variants.
+ *
+ * @version $Id$
+ * @since 1.2RC1
+ */
+@Component
+@Named("empty-objects-remover")
+@Singleton
+public class PatientEmptyObjectsRemover extends AbstractEventListener
+{
+    private static final EntityReference GENE_CLASS_REFERENCE = new EntityReference("GeneClass",
+        EntityType.DOCUMENT, Constants.CODE_SPACE_REFERENCE);
+
+    private static final EntityReference VARIANT_CLASS_REFERENCE = new EntityReference("GeneVariantClass",
+        EntityType.DOCUMENT, Constants.CODE_SPACE_REFERENCE);
+
+    private static final String GENE_KEY = "gene";
+
+    private static final String VARIANT_KEY = "cdna";
+
+    @Inject
+    private Execution execution;
+
+    /**
+     * Default constructor, sets up the listener name and the list of events to subscribe to.
+     */
+    public PatientEmptyObjectsRemover()
+    {
+        super("empty-objects-remover", new PatientChangingEvent());
+    }
+
+    @Override
+    public void onEvent(Event event, Object source, Object data)
+    {
+        XWikiDocument doc = (XWikiDocument) source;
+        XWikiContext context = (XWikiContext) this.execution.getContext().getProperty("xwikicontext");
+
+        Boolean changed = false;
+
+        Map<String, EntityReference> refs = new LinkedHashMap<String, EntityReference>();
+        refs.put(GENE_KEY, GENE_CLASS_REFERENCE);
+        refs.put(VARIANT_KEY, VARIANT_CLASS_REFERENCE);
+
+        for (String key : refs.keySet()) {
+            List<BaseObject> xWikiObjects = doc.getXObjects(refs.get(key));
+            if (xWikiObjects == null || xWikiObjects.isEmpty()) {
+                continue;
+            }
+            for (BaseObject object : xWikiObjects) {
+                if (object == null) {
+                    continue;
+                }
+                BaseStringProperty field = (BaseStringProperty) object.getField(key);
+                if (field == null || StringUtils.isEmpty(field.getValue())) {
+                    doc.removeXObject(object);
+                    changed = true;
+                }
+            }
+        }
+
+        try {
+            if (changed) {
+                context.getWiki().saveDocument(doc, "Removed empty object", true, context);
```

Comment:
@sdumitriu Is this the correct way of handling a PatientChangingEvent where data is modified?